enhancements/machine-api/spot-instances.md

```diff
@@ -0,0 +1,359 @@
+}
+```
+
+##### Azure
+
+###### Launching Instances
+
+To launch a VM as a Spot VM on Azure, the following 3 options need to be set within the [VirtualMachineProperties](https://github.com/Azure/azure-sdk-for-go/blob/8d7ac6eb6a149f992df6f0392eebf48544e2564a/services/compute/mgmt/2019-07-01/compute/models.go#L10274-L10309)
+when the instance is created:
+
+- Priority: This must be set to `Spot` to request a Spot VM
+
+- Eviction Policy: This has two options, `Deallocate` or `Delete`. Only `Deallocate` is valid when using Spot VMs and as such, this must be set to `Deallocate`.
+
+- BillingProfile (default: -1) : This is a struct containing a single field, `MaxPrice`.
+  This is a float representation of the maximum price the user wishes to pay for their VM.
+  This defaults to -1 which makes the maximum price the On-Demand price for the instance type.
+  This also means the instance will never be evicted for price reasons as Azure caps Spot Market prices at the On-Demand price.
+
+The only option that a user needs to interact with is the `MaxPrice` field within the `BillingProfile`, other fields only have 1 valid choice and as such can be inferred.
+Similar to AWS, we can make an optional struct for SpotVMOptions, which, if present, implies the priority is `Spot`.
+
+```
+type SpotVMOptions struct {
+  MaxPrice *float64 `json:”maxPrice,omitempty”`
+}
+
+type AzureMachineProviderSpec struct {
+  ...
+
+  SpotVMOptions *SpotVMOptions `json:”spotVMOptions,omitempty”`
+}
+```
+
+###### Deallocation
+Since Spot VMs are not deleted when they are preempted and instead are deallocated,
+we will need to monitor for deallocated instances and manually cleanup the leftover resources (VM, disks, networking).
+A MachineHealthCheck on Spot VMs should be capable of noticing that the machine has stopped and initiating a delete on the VM by the Machine actuator.
+
+### Risks and Mitigations
+
+## Design Details
+
+### Test Plan
+
+### Graduation Criteria
+
+#### Examples
+
+##### Dev Preview -> Tech Preview
+
+##### Tech Preview -> GA
+
+### Upgrade / Downgrade Strategy
+
+### Version Skew Strategy
+
+## Implementation History
+
+## Drawbacks
+
+## Alternatives
+
+Completely delegate the spot instance/feet behaviour to the provider in the fashion of [MachinePool](https://github.com/kubernetes-sigs/cluster-api/pull/1703).
+This would be a more disruptive change since the current OCP design expects a one to one relation between machine resources, provider instances and kubernetes nodes.
+This is particularly necessary for a kubelet to authenticate itself against the [machine approver](https://github.com/openshift/cluster-machine-approver) and join the cluster.
+
+## Future work
+
+### Termination Notices for GCP and Azure
+Each cloud provider provides a pre-warning when it is going to terminate an instance.
+In the first iteration of adding spot instance support, these will not be observed on GCP and Azure and workloads will be interrupted and rescheduled.
```

Comment:
This kind of seems like baseline functionality to me - why is it future work?  Is it really that hard?

Reply:
Originally we were going to not do this for any provider to limit the scope. But since we've decided to include termination for AWS in the first iteration, it probably does make sense to include a termination handler for the others too. For GCP this should be relatively straight forward as there are existing solutions we can learn from, for Azure, it will have to be all new work

WDYT @enxebre?

Reply:
I think we don't need the termination notice as a blocker for any provider.  There's plenty of workloads that people will want to run that can tolerate immediate termination.  With a properly managed group of spot instances (spread across types, and AZ'z) disruption should be minimal and would simulate a node going down every now and then.

I think for many people, 2 minute notification on AWS is not going to be sufficient to safely stop a lot of their stateful apps, so they're going to run into issues if they schedule those workloads onto the spot instances.

It's definitely nice to have, but I don't feel it's critical to successfully consuming spot instances.